feat: tracks

is-it-ayush opened the pull request:
### Issue
closes #305 ...most of it except the following,
- probably should appear on user profile (only to the logged in user)

### Changes

there are a lot. i suggest you to go through the review instead.

### Known Bugs
- `<TrackCard />`'s `key={}` problem. i'm not sure how to debug this one. i spent some time but couldn't really figure out what 
- logging out doesn't refresh the `/tracks` page. it should refresh the data which should remove the `Your Tracks` section, although i'm unsure why it doesn't.
- width issue when `hover:scale-110` on challenges over at `/tracks`. i have css skill issue.
- enroll state doesn't update after you  enroll/unenroll on `/tracks/[slug]` page. 

### SC
[Screencast from 2023-08-27 03-44-42.webm](https://github.com/bautistaaa/typehero/assets/36449128/1e138f7b-eabf-4f9b-86d4-e0c11147b303)

bautistaaa commented:
initial thoughts.

- we're only going to have like 5 tracks for a while(we are curating those) so i dont think bucketing the tracks into separate carousel makes sense. at least for the initial release. i do think eventually we'll need this
- need to make the carousel smart. it shouldn't always display arrows(out of scope)
- i dont think we should should show the enrollment number(im assuming thats what that is).

i can help address these in addition to the bugs listed in the PR description but

bautistaaa commented on apps/web/src/app/explore/_components/explore-section.tsx:
```diff
@@ -1,21 +1,33 @@
+  HARD: 3,
+  EXTREME: 4,
+};
+
```
```suggestion
const difficultyToNumber = {
  BEGINNER: 0,
  EASY: 1,
  MEDIUM: 2,
  HARD: 3,
  EXTREME: 4,
} as const;
```

is-it-ayush replied:
done!

bautistaaa commented on apps/web/src/app/explore/_components/explore-section.tsx:
```diff
@@ -28,7 +40,7 @@ const TITLES_BY_DIFFICULTY = {
 };
 
-const COLORS_BY_DIFFICULTY = {
+export const COLORS_BY_TAGS = {
```
as const this

is-it-ayush replied:
re solved

bautistaaa commented on apps/web/src/app/explore/_components/view-more-button.tsx:
```diff
@@ -3,9 +3,8 @@
-import type { ExploreChallengeData } from './explore.action';
 
-const BUTTONS_BY_DIFFICULTY = {
+export const BUTTON_BY_TAGS = {
```
as const this

is-it-ayush replied:
resolved!

bautistaaa commented on apps/web/src/app/tracks/_components/enroll-button.tsx:
```diff
@@ -0,0 +1,20 @@
+  return (
+    <Button
+      onClick={async () => {
+        await action();
```
error handling?

is-it-ayush replied:
oh yep, i actually had error handling in the `track-details.tsx` because i wanted custom messages. Although u give me an idea, wait i'll test it out. I was stupid & didn't think of this before. 💀

is-it-ayush replied:
RESOLVEDEDD!!!

bautistaaa commented on apps/web/src/app/tracks/_components/index.tsx:
```diff
@@ -1,61 +1,114 @@
+          tracks={enrolledTracks}
+        />
+      )}
+      <PopularTrackSection
```
lets just list them out vs carousel. we're launching with small number of tracks

is-it-ayush replied:
oki! since the carousel is generalised. you can just plop it the children at any time you need it in future. (which u will as [you've said here trash](https://github.com/bautistaaa/typehero/pull/451#issuecomment-1694538323))

bautistaaa commented on apps/web/src/app/tracks/_components/index.tsx:
```diff
@@ -1,61 +1,114 @@
+ * Fetches popular tracks. Popularity is determined by number of user's enrolled in it.
+ * todo: change this behavior for now it's fine.
+ */
+function getPopularTracks() {
```
probably dont need this and can return them alphabetically or other criteria

bautistaaa commented on apps/web/src/app/tracks/_components/track-challenge-card.tsx:
```diff
@@ -0,0 +1,77 @@
+  HARD: 'to-red-600/20 dark:to-red-300/20',
+  EXTREME: 'to-orange-600/20 dark:to-orange-300/20',
+};
+
```
as const the two enums above

is-it-ayush replied:
done! : D

bautistaaa commented on apps/web/src/app/tracks/_components/track-details.tsx:
```diff
@@ -0,0 +1,112 @@
+  // things you need to do to skip first render & call server action on client component.
+  // this is nightmare fuel.
+  // todo: find better way
+  useEffect(() => {
```
why cant this be fetched in the RSC and fed in?

is-it-ayush replied:
i simplified it a lot trash, we don't have this mess anymore. : D

is-it-ayush commented:
yo @bautistaaa!

> we're only going to have like 5 tracks for a while(we are curating those) so i dont think bucketing the tracks into separate carousel makes sense. at least for the initial release. i do think eventually we'll need this

we already have the carousel component in which we can put `{children}` in. so it should be good to use anytime you need it in future.

> need to make the carousel smart. it shouldn't always display arrows(out of scope)

yep, this could be separate issue trash!!

> i dont think we should should show the enrollment number(im assuming thats what that is).

i removed it for now (commented it out). if you wanna show any statistics related to tracks in future, just uncomment it and remove all nested nodes.

> i can help address these in addition to the bugs listed in the PR description but

i fixed the rsc mess & the enroll state bug. we still have 3 bugs to go. here is an updated bug list. 
- [ ] `<TrackCard />`'s `key={}` problem. i spent time on it but i'm not sure how to fix this one.
- [ ] logging out doesn't refresh the `/tracks` page. logging out doesn't reload the page which it should since it has `window.location.reload()` on it.
- [ ] width issue when hover:scale-110 on challenges over at /tracks. i have css skill issue.
